File: 0199.BinaryTreeRightSideView.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 func rightSideView(root *TreeNode) []int {
    ans := []int{}
    dfs(root, 0, &ans)
    /* BFS
    if root == nil {
        return nil
    }
    ans := []int{}
    queue := []*TreeNode{root}
    for len(queue) > 0 {
        ans = append(ans, queue[len(queue)-1].Val)
        nxt := []*TreeNode{}
        for _, node := range queue {
            if node.Left != nil {
                nxt = append(nxt, node.Left)
            }
            if node.Right != nil {
                nxt = append(nxt, node.Right)
            }
        }
        queue = nxt
    } */
    return ans
}

func dfs(node *TreeNode, depth int, ans *[]int) {
    if node == nil {
        return
    }
    if len(*ans) == depth {
        *ans = append(*ans, node.Val)
    }
    dfs(node.Right, depth+1, ans)
    dfs(node.Left, depth+1, ans)
}

File: 0000-0999/0199.BinaryTreeRightSideView.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def rightSideView(self, root):
        """
        :type root: TreeNode
        :rtype: List[int]
        """
        if not root:
            return []
        '''
        ans = []
        currLine = [root]
        nextLine = []
        while currLine:
            ans.append(currLine[-1].val)
            while currLine:
                node = currLine.pop(0)
                if node.left:
                    nextLine.append(node.left)
                if node.right:
                    nextLine.append(node.right)
            currLine, nextLine = nextLine, currLine
        return ans
        '''
        r = self.rightSideView(root.right)
        l = self.rightSideView(root.left)
        return [root.val] + r + l[len(r):]
        